src/osemgrep/cli_scan/Core_runner.ml

```diff
@@ -209,6 +210,199 @@ let pp_status rules targets lang_jobs respect_git_ignore ppf () =
          []
     |> List.rev)
 
+let errors_to_skipped (errors : Out.core_error list) : Out.skipped_target list =
+  errors
+  |> Common.map (fun Out.{ location; message; rule_id; _ } ->
+         Out.
+           {
+             path = location.path;
+             reason = Analysis_failed_parser_or_internal_error;
+             details = message;
+             rule_id;
+           })
+
+let analyze_skipped (skipped : Out.skipped_target list) =
+  let reason_ht = Hashtbl.create 13 in
+  List.iter
+    (fun r -> Hashtbl.add reason_ht r [])
+    Out.
+      [
+        Semgrepignore_patterns_match;
+        Cli_include_flags_do_not_match;
+        Cli_exclude_flags_match;
+        Exceeded_size_limit;
+      ];
+  let skipped_by_reason (Out.{ reason; _ } as e : Out.skipped_target) =
+    let reason =
+      match reason with
+      | Out.Gitignore_patterns_match
+      | Semgrepignore_patterns_match ->
+          Out.Semgrepignore_patterns_match
+      | Too_big
+      | Exceeded_size_limit ->
+          Out.Exceeded_size_limit
+      | Cli_include_flags_do_not_match -> Out.Cli_include_flags_do_not_match
+      | Cli_exclude_flags_match -> Out.Cli_exclude_flags_match
+      | Always_skipped
+      | Analysis_failed_parser_or_internal_error
+      | Excluded_by_config
+      | Wrong_language
+      | Minified
+      | Binary
+      | Irrelevant_rule
+      | Too_many_matches ->
+          assert false
+    in
+    let v = e :: Option.value ~default:[] (Hashtbl.find_opt reason_ht reason) in
+    Hashtbl.replace reason_ht reason v
+  in
+  List.iter skipped_by_reason skipped;
+  ( (try Hashtbl.find reason_ht Out.Semgrepignore_patterns_match with
+    | Not_found -> []),
+    (try Hashtbl.find reason_ht Out.Cli_include_flags_do_not_match with
+    | Not_found -> []),
+    (try Hashtbl.find reason_ht Out.Cli_exclude_flags_match with
+    | Not_found -> []),
+    try Hashtbl.find reason_ht Out.Exceeded_size_limit with
+    | Not_found -> [] )
```

Comment:
If you don't mind getting a list of (key, value list) pairs rather than a tuple, we have a `Common.group_by`* functions that can do such clustering.

Reply:
thanks for the pointer, now using that in [0a881eb](https://github.com/returntocorp/semgrep/pull/7456/commits/0a881eb7d981b569b0abee0b29d24e71869e26ba)